caddyhttp/httpserver/middleware.go

```diff
@@ -48,20 +47,48 @@ type (
 
 	// RequestMatcher checks to see if current request should be handled
 	// by underlying handler.
-	//
-	// TODO The long term plan is to get all middleware implement this
-	// interface and have validation done before requests are dispatched
-	// to each middleware.
 	RequestMatcher interface {
 		Match(r *http.Request) bool
 	}
+
+	// Config is a middleware configuration.
+	// This makes it possible for middlewares to have a common
+	// configuration interface.
+	//
+	// TODO The long term plan is to get all middleware implement this
+	// interface for configurations.
+	Config interface {
+		RequestMatcher
+		BasePath() string
+	}
+
+	// ConfigSelector selects a configuration.
+	//
+	// TODO The long term plan is to get all middleware to use this.
+	ConfigSelector []Config
 )
 
 // ServeHTTP implements the Handler interface.
 func (f HandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 	return f(w, r)
 }
 
+// Select selects a Config.
+// This chooses the config with the longest length.
+//
+// TODO A better solution.
+func (c ConfigSelector) Select(r *http.Request) (config Config) {
+	for i := range c {
+		if !c[i].Match(r) {
+			continue
+		}
+		if config == nil || len(c[i].BasePath()) > len(config.BasePath()) {
```

Comment:
This is the only place that BasePath() is used; I can see why it is, but I wonder if, instead of a Config interface as you have it here, maybe a new interface that compares whether one match is superior to another. In fact, this can be reduced to a sorting problem! And there's a standard `sort` package. What do you think about that train of thought?